python/lsst/daf/butler/core/logging.py

```diff
@@ -21,11 +21,51 @@
 
 from __future__ import annotations
 
-__all__ = ("VERBOSE",)
+__all__ = ("VERBOSE", "ButlerMDC")
 
 import logging
 
 VERBOSE = (logging.INFO + logging.DEBUG) // 2
 """Verbose log level"""
 
 logging.addLevelName(VERBOSE, "VERBOSE")
+
+
+class MDCDict(dict):
+    """Dictionary for MDC data.
+
+    This is internal class used for better formatting of MDC in Python logging
+    output. It behaves like `defaultdict(str)` but overrides ``__str__`` and
+    ``__repr__`` method to produce output better suited for logging records.
+    """
+
+    def __getitem__(self, name: str) -> str:
+        """Return value for a given key or empty string for missing key.
+        """
+        return self.get(name, "")
+
+    def __str__(self) -> str:
+        """Return string representation, strings are interpolated without
+        quotes.
+        """
+        items = (f"{k}={self[k]}" for k in sorted(self))
+        return "{" + ", ".join(items) + "}"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+class ButlerMDC:
+    """Handle setting and unsetting of global MDC records."""
+
+    _MDC = MDCDict()
```

Comment:
One potential inconsistency is that `log4cxx` MDC is per-thread, while this one is global. I don't think we ever used that per-thread feature on Python side in LSST code (the only place that I know of is C++ code in qserv). If it is ever needed we may need to change this to use thread-local storage (which opens another can of worms that was in some way solved in `lsst.log`).

Reply:
Python logging itself does handle threading but I had not pondered the situation of multiple threads setting their own MDC. We haven't got any code that wants to do that so I think I'll punt for now but will add a comment about it.